### PR TITLE
feat(pipenv): Use `@renovatebot/detect-tools` for constraints detection

### LINKED ISSUE #29000: Move tool detection to renovatebot/detect-tools
### Describe the proposed change(s).

We should move tool detection to a separate repo/module so that it can be shared outside Renovate, particularly with some Containerbase tooling we are building.

Start with Poetry and Pipenv.

Example from: https://github.com/renovatebot/renovate/blob/main/lib/modules/manager/pipenv/artifacts.ts

Instead of a `getPythonConstraint()` function where we pass _content_ to the function, we should instead call our new external library and pass it a fully qualified _path_/directory.

### PR DESCRIPTION
**Pull request by zharinov**
## Changes

Use drop-in library replacement for pypi constraint detection.

## Context

- Closes: #29000
- https://github.com/renovatebot/detect-tools
- https://www.npmjs.com/package/@renovatebot/detect-tools?activeTab=code

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work (please select one)

I have verified these changes via:

- [ ] Code inspection only, or
- [x] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [ ] Both unit tests + ran on a real repository

### CODE REVIEW

File: lib/modules/manager/pipenv/artifacts.ts [hunk 1]
```diff
@@ -1,5 +1,5 @@
+import * as detectTools from '@renovatebot/detect-tools';
```
Comment (viceice) [inline lib/modules/manager/pipenv/artifacts.ts:1]: why the import star is required? can't we use a default export?

**Review by viceice — COMMENTED** [review]
new code looks promising, please also let me review changes to the detect tools. I need them later for containerbase.

File: lib/modules/manager/pipenv/schema.ts
Comment (viceice) [inline lib/modules/manager/pipenv/schema.ts:1]: Please don't delete, it should be used to extract locked version later

File: lib/modules/manager/pipenv/artifacts.ts [hunk 2]
```diff
@@ -289,16 +139,11 @@ export async function updateArtifacts({
...
-      newPipfileContent,
-      existingLockFileContent,
-      config,
-    );
-    const pipenvConstraint = getPipenvConstraint(
-      existingLockFileContent,
-      config,
-    );
+    const pipfileDir = getParentDir(ensureLocalPath(pipfileName));
+    const tagConstraint = await pipenvDetect.getPythonConstraint(pipfileDir);
```
Comment (viceice) [inline lib/modules/manager/pipenv/artifacts.ts:143]:
```suggestion
    const tagConstraint = config.constraints?.python ?? await pipenvDetect.getPythonConstraint(pipfileDir);
```
we should allow configure python version via constraints too

File: lib/modules/manager/pipenv/artifacts.ts [hunk 3]
```diff
@@ -289,16 +139,11 @@ export async function updateArtifacts({
...
+    const pipenvConstraint =
+      config?.constraints?.pipenv ??
```
Comment (viceice) [inline lib/modules/manager/pipenv/artifacts.ts:145]:
```suggestion
      config.constraints?.pipenv ??
```

File: lib/modules/manager/pipenv/artifacts.spec.ts [hunk 4]
```diff
@@ -845,19 +880,22 @@ describe('modules/manager/pipenv/artifacts', () => {
...
 
   it('uses pipenv version from config', async () => {
     GlobalConfig.set(dockerAdminConfig);
+    fsExtra.stat.mockResolvedValueOnce({} as never);
     fsExtra.ensureDir.mockResolvedValue(undefined as never);
 
     const oldLock = JSON.stringify({
       default: { pipenv: { version: '==2020.8.13' } },
-    } satisfies PipfileLockSchema) as never;
+    }) as never;
```
Comment (viceice) [inline lib/modules/manager/pipenv/artifacts.spec.ts:898]: revert

File: lib/modules/manager/pipenv/artifacts.spec.ts [hunk 5]
```diff
@@ -24,7 +24,6 @@ import {
   extractEnvironmentVariableName,
   getMatchingHostRule,
 } from './artifacts';
-import type { PipfileLockSchema } from './schema';
```
Comment (viceice) [inline lib/modules/manager/pipenv/artifacts.spec.ts:27]: revert, we should still use it see other comments

File: lib/modules/manager/pipenv/artifacts.ts (hunk 2 above)
Comment (zharinov) [reply lib/modules/manager/pipenv/artifacts.ts:143]: Yep, I don't remember why I removed it, seems like by error

File: lib/modules/manager/pipenv/artifacts.spec.ts (hunk 5 above)
Comment (zharinov) [reply lib/modules/manager/pipenv/artifacts.spec.ts:27]: It doesn't exist anymore

Comment (zharinov) [reply lib/modules/manager/pipenv/artifacts.spec.ts:27]: It was heavily refactored in the `detect-tools` package

File: lib/modules/manager/pipenv/artifacts.ts [hunk 6]
```diff
@@ -289,16 +139,13 @@ export async function updateArtifacts({
...
-      existingLockFileContent,
-      config,
-    );
-    const pipenvConstraint = getPipenvConstraint(
-      existingLockFileContent,
-      config,
-    );
+    const pipfileDir = getParentDir(ensureLocalPath(pipfileName));
+    const tagConstraint =
+      config?.constraints?.python ??
```
Comment (viceice) [inline lib/modules/manager/pipenv/artifacts.ts:144]:
```suggestion
      config.constraints?.python ??
```
config is never null or undefined here

File: lib/modules/manager/pipenv/artifacts.ts [hunk 7]
```diff
@@ -289,16 +139,13 @@ export async function updateArtifacts({
...
+      (await pipenvDetect.getPythonConstraint(pipfileDir));
+    const pipenvConstraint =
+      config?.constraints?.pipenv ??
```
Comment (viceice) [inline lib/modules/manager/pipenv/artifacts.ts:147]:
```suggestion
      config.constraints?.pipenv ??
```

**Review by viceice — COMMENTED** [review]
https://github.com/renovatebot/renovate/blob/607d7e3e1bc973a65a9a964fbdd7d050dc3a6239/lib/modules/manager/pipenv/extract.ts#L171

this should be replaced too

File: lib/modules/manager/pipenv/extract.spec.ts [hunk 8]
```diff
@@ -99,7 +114,7 @@ describe('modules/manager/pipenv/extract', () => {
           },
         ],
         extractedConstraints: {
-          python: '== 3.6.*',
+          python: '== 3.6.2',
```
Comment (viceice) [inline lib/modules/manager/pipenv/extract.spec.ts:117]: why this change? 🤔

**Review by viceice — COMMENTED** [review]
otherwise LGTM

File: lib/modules/manager/pipenv/extract.spec.ts (hunk 8 above)
Comment (zharinov) [reply lib/modules/manager/pipenv/extract.spec.ts:117]: This is how artifact extraction logic works, it turned out to be different than `extract.ts`

**Review by viceice — APPROVED** [review]
